meta-mbl-distro/classes/mbl-partitions.bbclass

```diff
@@ -24,138 +24,405 @@ inherit mbl-artifact-names
+                prev_offset = offset
+                prev_size = self.size_KiB
+
+        def _set_offset_var(self, offset_var, prev_offset, prev_size):
+            offset = None
+            have_offset = self._is_var_set(offset_var)
+            no_default = self._get_bool_var("MBL_{}_NO_DEFAULT_OFFSET".format(self.name))
+            have_prev = prev_offset is not None and prev_size is not None
+            if have_offset:
+                offset = self._get_size_var_or_fatal(offset_var)
+                self.offsets_KiB.append(offset)
+                if offset % self.align_KiB:
+                    bb.fatal("{}={}KiB is badly aligned. Expected alignment is {}KiB".format(
+                        offset_var, offset, self.align_KiB
+                    ))
+                if have_prev and prev_offset + prev_size > offset:
+                    bb.fatal("{}={}KiB overlaps previous partition with offset {}KiB and size {}KiB".format(
+                        offset_var, offset, prev_offset, prev_size
+                    ))
+            elif no_default:
+                bb.fatal("{} is not specified".format(offset_var))
+            elif have_prev:
+                offset = self._calculate_next_offset(prev_offset, prev_size)
+                self.d.setVar(offset_var, str(offset))
+                self.offsets_KiB.append(offset)
+            else:
+                bb.fatal("{} is not set", offset_var)
+            return offset
+
+        def _calculate_next_offset(self, prev_offset, prev_size):
+            last_byte_of_prev = prev_offset + prev_size - 1
+            return last_byte_of_prev + self.align_KiB - (last_byte_of_prev % self.align_KiB)
+
+        def _get_size_var_or_fatal(self, var):
+            var_str = self.d.getVar(var, True)
+            if var_str is None:
+                bb.fatal("{} has not been set".format(var))
+                return 0
+            if not var_str.isdigit():
+                bb.fatal("{} (\"{}\") must contain only decimal digits".format(var, var_str))
+                return 0
+            return int(var_str)
+
+        def _is_var_set(self, var):
+            return bool(self.d.getVar(var, True))
+
+        def _get_bool_var(self, var):
+            val = self.d.getVar(var, True)
+            return val is not None and bool(int(val))
+
+        def __str__(self):
+            if self.skip:
+                return "{}: SKIP".format(self.name)
+
+            return "{}: FS={}, banked={} align={}KiB size={}KiB offsets={}".format(
+                self.name,
+                self.is_fs,
+                self.is_banked,
+                self.align_KiB,
+                self.size_KiB,
+                ["{}KiB".format(offset) for offset in self.offsets_KiB]
+            )
+
+    part_names = d.getVar("MBL_PARTITION_NAMES", True).split()
+    parts = []
+    prev_part = None
+    for part_name in part_names:
+        part = MaybeBankedPartition(d, part_name, prev_part)
```

Comment:
This seems strange; you pass an instance of `MaybeBankedPartition` in as an argument to another instance of `MaybeBankedPartition` so the object has knowledge of the previous partition size and offset right? Does it make sense to have another `Partitions` class which contains a list of `MaybeBankedPartition` classes, and have the `Partitions` parent class do the alignment and offset fiddling? I think the approach you've chosen here makes the code harder to understand in general.

Reply:
I sort of agree... I think the class is conflating several things:
1. Validation of BitBake variable values.
2. Setting/calculating BitBake variable values when they haven't been set explicitly by e.g. a MACHINE config.
3. Providing an interface to access partition information.

I'll see if I can come up with something better...

Reply:
I've got rid of the class in https://github.com/ARMmbed/meta-mbl/pull/617/commits/34e632c489ddce4f816e5f19992c7b47b7c6432a. The methods have been refactored and are now free functions. The data for each partition is now just a `dict`. I'm still not convinced I've got it right... Let me know what you think.

Reply:
I think it's an improvement. I probably would've preferred to see a the partition data as a class with public attributes rather than a dict. Don't think it's that important though.